src/main/java/org/jitsi/videobridge/cc/vp8/VP8AdaptiveTrackProjectionContext.java

```diff
@@ -45,42 +42,16 @@
     private final Logger logger;
 
     /**
-     * The time (in clock rate samples) to wait before discarding a
-     * non-fully-projected frame (see {@link #cleanupFrameProjectionMap}).
+     * The time series logger for this instance.
      */
-    private final long WAIT_SAMPLES = 45000;
+    private static final TimeSeriesLogger timeSeriesLogger
+        = TimeSeriesLogger.getTimeSeriesLogger(VP8AdaptiveTrackProjectionContext.class);
 
     /**
-     * A map of partially transmitted {@link VP8FrameProjection}s, i.e.
-     * projections of VP8 frames for which we haven't transmitted all their
-     * packets.
-     *
-     * Fully transmitted and skipped frames are removed from the map for
-     * housekeeping purposes, i.e. to prevent the map from growing too big.
-     *
-     * The purpose of this map is to enable forwarding and translation of
-     * recovered packets of partially transmitted frames and partially
-     * transmitted frames _only_.
-     *
-     * Recovered packets of fully transmitted frames (this can happen for
-     * example when the sending endpoint probes for bandwidth with duplicate
-     * packets over the RTX stream) are dropped as they're not needed anymore.
+     * A map that stores the per-encoding VP8 frame maps.
      */
-    private final SortedMap<Long, VP8FrameProjection>
-        vp8FrameProjectionMap = new ConcurrentSkipListMap<>(
-            /* This is only a valid Comparitor if timestamp diffs of all
-             * timestamps are within half the number space.  (This property is
-             * assured by #cleanupFrameProjectionMap.)
-             */
-            (t1, t2) -> (int)RtpUtils.getTimestampDiff(t1, t2));
-
-    /**
-     * A map that stores the maximum sequence number of frames that are not
-     * (yet) accepted/projected. The map goes from ssrc -> timestamp -> highest
-     * sequence number.
-     */
-    private final Map<Long, LRUCache<Long, Integer>>
-        ssrcToFrameToMaxSequenceNumberMap = new HashMap<>();
+    private final Map<Long, VP8FrameMap>
+        vp8FrameMaps = new HashMap<>();
```

Comment:
I'm still going through the code, so apologies for the potentially premature question; do we really need to remember frame projections from other ssrcs? isn't it sufficient to clean the map every time we switch simulcast streams (which refreshes the decoder as it can only happen on a keyframe, so we don't care if we leave undecodable frames at the receiver)

Reply:
If we've already received (and decided not to project) TL0s after a keyframe, we need to make sure that we don't subsequently project that keyframe, because the video will get stuck.

I wanted to keep the per-frame route-or-don't logic centralized in `VP8QualityFilter`, across all ssrcs, which means we need to maintain per-frame state for all the ssrcs.

Reply:
so you're talking about the scenario where we're forwarding stream_{A}, reject TL0_{n+1} of stream_{B} and then switch to stream_{B} upon receipt of keyframe_{n} of stream_{B}? Am I understanding correctly?

Reply:
Right, exactly.  I have logic that checks for that case (line 297 of VP8AdaptiveTrackProjectionContext).